tests/test_core_attention.py

```diff
@@ -109,3 +109,52 @@ def test_amp_attention_sparsecs(device):
 
     expected_device = torch.float32
     assert r.dtype == expected_device
+
+
+@pytest.mark.parametrize("device", ["cuda"])
+@pytest.mark.parametrize("data_type", [torch.float16, torch.float32])
+def test_switch_blocksparse(device, data_type):
+    b, s, d = 8, 128, 32
+
+    a = torch.rand(b, s, d, device=device, dtype=data_type)
+
+    # custom causal mask
+    m_custom = torch.triu(
+        torch.ones(s, s, device=device, dtype=a.dtype) * float("-inf"), diagonal=1
+    )
+    m_custom_bool = m_custom != float("-inf")
+    m_sparse = SparseCS(m_custom_bool, device)
+    # causal mask with causal flag
+    m_att_mask = AttentionMask.make_causal(s, s, device, dtype=a.dtype)
+
+    # Check that a switch to blocksparse is only triggered by causal flag
+    with torch.cuda.amp.autocast():
+        r_custom = scaled_dot_product_attention(a, a, a, m_custom)
+        r_sparse = scaled_dot_product_attention(a, a, a, m_sparse)
+        r_att_mask = scaled_dot_product_attention(a, a, a, m_att_mask)
+
+    r_att_mask = r_att_mask.squeeze()
+
+    expected_device = torch.float32
+    assert r_sparse.dtype == expected_device
+
+    if r_custom.dtype == r_att_mask.dtype:
+        assert torch.allclose(r_custom, r_att_mask, atol=1e-6, rtol=1e-3)
```

Comment:
@blefaudeux @fmassa @dianaml0 In this test I had to increase the tolerances of `torch.allclose()` from the default `atol=1e-8` and `rtol=1e-5` pretty significantly for the assert to pass. Something similar, `assert_almost_equal()`, from is used [here](https://github.com/facebookresearch/xformers/blob/main/tests/test_triton_blocksparse.py#L204) to test parity between standard SDP attention and blocksparse attention.

Is this difference between SDP attention and blocksparse acceptable in this situation? I.e., we're silently switching to blocksparse as of now, but should we inform the user? Or are there any other steps I should take in the code

Reply:
we can force pytorch to use tf32 in that case also, else I think that it's fine if the tolerance relaxation is limited to this case ?

Reply:
@fmassa @dianaml0 Just pushed what I currently have as my benchmark file. It's a WIP so little messy, but its there in case you guys want to take a look.